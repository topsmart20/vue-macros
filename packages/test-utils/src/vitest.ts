import path from 'node:path'
import glob, { type Options as GlobOptions } from 'fast-glob'
import { describe, expect, test } from 'vitest'
import { normalizePath } from '@rollup/pluginutils'

interface Options {
  params?: [name: string, values?: any[]][]
  promise?: boolean
}

const SKIP_VUE2 = !!process.env.SKIP_VUE2

export async function testFixtures(
  globs: string | string[],
  exec: (args: Record<string, any>, id: string) => any,
  options?: GlobOptions & Options,
): Promise<void>
export async function testFixtures(
  files: Record<string, string>,
  exec: (args: Record<string, any>, id: string, code: string) => any,
  options?: Options,
): Promise<void>
export async function testFixtures(
  globsOrFiles: string | string[] | Record<string, string>,
  cb: (args: Record<string, any>, id: string, code: string) => any,
  { params, promise, ...globOptions }: GlobOptions & Options = {},
) {
  let files: Record<string, string | undefined>

  if (typeof globsOrFiles === 'string' || Array.isArray(globsOrFiles)) {
    files = Object.fromEntries(
      (await glob(globsOrFiles, globOptions)).map((file) => [file, undefined]),
    )
  } else {
    files = globsOrFiles
  }

  for (const [id, code] of Object.entries(files)) {
    makeTests(id, code, [[normalizePath(id)], ...(params || [])])()
  }

  function getName(name: string, value: any) {
    return value !== undefined ? `${name} = ${String(value)}` : name
  }

  function makeTests(
    id: string,
    code: string | undefined,
    params: NonNullable<Options['params']>,
    args: Record<string, any> = {},
  ) {
    const [currParams, ...restParams] = params
    const [name, values = [undefined]] = currParams
    if (restParams.length > 0) {
      return () => {
        for (const value of values) {
          const currArgs = { ...args, [name]: value }

          describe(
            getName(name, value),
            makeTests(id, code, restParams, currArgs),
          )
        }
      }
    } else {
      return () => {
        for (const value of values) {
          const testName = getName(name, value)
          const isSkip = testName.includes('vue2') && SKIP_VUE2

          test.skipIf(isSkip)(testName, async () => {
            const currArgs = { ...args, [name]: value }
            const execute = () =>
              cb(currArgs, path.resolve(globOptions.cwd || '/', id), code!)
            if (id.includes('error')) {
              if (promise) {
                await expect(execute()).rejects.toThrowErrorMatchingSnapshot()
              } else {
                expect(execute).toThrowErrorMatchingSnapshot()
              }
            } else if (promise) {
              await expect(
                (execute() as Promise<any>).catch((error) => {
                  console.warn(error)
                  return Promise.reject(error)
                }),
              ).resolves.toMatchSnapshot()
            } else {
              expect(execute()).toMatchSnapshot()
            }
          })
        }
      }
    }
  }
}
