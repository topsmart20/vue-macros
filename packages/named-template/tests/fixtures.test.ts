import { resolve } from 'node:path'
import { describe } from 'vitest'
import {
  RollupEsbuildPlugin,
  RollupRemoveVueFilePathPlugin,
  RollupVue,
  RollupVueJsx,
  rollupBuild,
  testFixtures,
} from '@vue-macros/test-utils'
import VueNamedTemplate from '../src/rollup'

describe('fixtures', async () => {
  await testFixtures(
    'tests/fixtures/*.vue',
    (args, id) =>
      rollupBuild(id, [
        VueNamedTemplate(),
        RollupVue(),
        RollupVueJsx(),
        RollupRemoveVueFilePathPlugin(),
        RollupEsbuildPlugin({
          target: 'esnext',
        }),
      ]),
    {
      cwd: resolve(__dirname, '..'),
      promise: true,
    }
  )
})
