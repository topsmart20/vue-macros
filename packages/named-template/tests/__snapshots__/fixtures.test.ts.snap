// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fixtures > tests/fixtures/basic.vue 1`] = `
"import { defineComponent, createElementVNode, openBlock, createElementBlock, Fragment, toDisplayString, createTextVNode, createBlock } from 'vue';
import _export_sfc from '/plugin-vue/export-helper';

const foo = \\"foo\\";
var _sfc_main = /* @__PURE__ */ defineComponent({
  __name: \\"basic\\",
  setup(__props, { expose: __expose }) {
    __expose();
    const __returned__ = { foo };
    Object.defineProperty(__returned__, \\"__isScriptSetup\\", { enumerable: false, value: true });
    return __returned__;
  }
});

const _hoisted_1$1 = /*#__PURE__*/createElementVNode(\\"div\\", null, \\"Test\\", -1 /* HOISTED */);

function render$1(_ctx, _cache, $props, $setup, $data, $options) {
  return (openBlock(), createElementBlock(Fragment, null, [
    _hoisted_1$1,
    createElementVNode(\\"span\\", null, toDisplayString($setup.foo))
  ], 64 /* STABLE_FRAGMENT */))
}

var block0 = {
  render: (...args) => {
    const r = render$1(...args);
    return typeof r === 'string' ? createTextVNode(r) : r
  }
};

const _hoisted_1 = /* @__PURE__ */ createElementVNode(
  \\"span\\",
  null,
  \\"hello\\",
  -1
  /* HOISTED */
);
function render(...args) {
 let [_ctx, _cache, $props, $setup, $data, $options] = args;
  return openBlock(), createElementBlock(
    Fragment,
    null,
    [
      _hoisted_1,
      (openBlock(), createBlock(Fragment, null, [block0.render(...args)]))
    ],
    64
    /* STABLE_FRAGMENT */
  );
}

if (typeof block0 === \\"function\\")
  block0(_sfc_main);
var basic = /* @__PURE__ */ _export_sfc(_sfc_main, [[\\"render\\", render], __FILE__]);

export { basic as default };
"
`;

exports[`fixtures > tests/fixtures/dash-name.vue 1`] = `
"import { defineComponent, openBlock, createElementBlock, renderSlot, createTextVNode, createBlock, Fragment } from 'vue';
import _export_sfc from '/plugin-vue/export-helper';

var _sfc_main = /* @__PURE__ */ defineComponent({
  __name: \\"dash-name\\",
  setup(__props, { expose: __expose }) {
    __expose();
    const __returned__ = {};
    Object.defineProperty(__returned__, \\"__isScriptSetup\\", { enumerable: false, value: true });
    return __returned__;
  }
});

const _hoisted_1$1 = { class: \\"btn-content\\" };

function render$1(_ctx, _cache, $props, $setup, $data, $options) {
  return (openBlock(), createElementBlock(\\"div\\", _hoisted_1$1, [
    renderSlot(_ctx.$slots, \\"default\\")
  ]))
}

var block0 = {
  render: (...args) => {
    const r = render$1(...args);
    return typeof r === 'string' ? createTextVNode(r) : r
  }
};

const _hoisted_1 = { class: \\"a-btn\\" };
function render(...args) {
 let [_ctx, _cache, $props, $setup, $data, $options] = args;
  return openBlock(), createElementBlock(\\"div\\", _hoisted_1, [
    (openBlock(), createBlock(Fragment, null, [block0.render(...args)]))
  ]);
}

if (typeof block0 === \\"function\\")
  block0(_sfc_main);
var dashName = /* @__PURE__ */ _export_sfc(_sfc_main, [[\\"render\\", render], __FILE__]);

export { dashName as default };
"
`;

exports[`fixtures > tests/fixtures/multi.vue 1`] = `
"import { defineComponent, openBlock, createElementBlock, createTextVNode, createElementVNode, Fragment, createBlock, createCommentVNode } from 'vue';
import _export_sfc from '/plugin-vue/export-helper';

const pager = \\"top\\";
var _sfc_main = /* @__PURE__ */ defineComponent({
  __name: \\"multi\\",
  setup(__props, { expose: __expose }) {
    __expose();
    const __returned__ = { pager };
    Object.defineProperty(__returned__, \\"__isScriptSetup\\", { enumerable: false, value: true });
    return __returned__;
  }
});

function render$1(_ctx, _cache, $props, $setup, $data, $options) {
  return (openBlock(), createElementBlock(\\"span\\", null, \\"This is pager\\"))
}

var block0 = {
  render: (...args) => {
    const r = render$1(...args);
    return typeof r === 'string' ? createTextVNode(r) : r
  }
};

const _hoisted_1 = /* @__PURE__ */ createElementVNode(
  \\"span\\",
  null,
  \\"Here is data table\\",
  -1
  /* HOISTED */
);
function render(...args) {
 let [_ctx, _cache, $props, $setup, $data, $options] = args;
  return openBlock(), createElementBlock(
    Fragment,
    null,
    [
      $setup.pager === \\"top\\" ? (openBlock(), createBlock(Fragment, { key: 0 }, [block0.render(...args)])) : createCommentVNode(\\"v-if\\", true),
      _hoisted_1,
      $setup.pager === \\"bottom\\" ? (openBlock(), createBlock(Fragment, { key: 1 }, [block0.render(...args)])) : createCommentVNode(\\"v-if\\", true)
    ],
    64
    /* STABLE_FRAGMENT */
  );
}

if (typeof block0 === \\"function\\")
  block0(_sfc_main);
var multi = /* @__PURE__ */ _export_sfc(_sfc_main, [[\\"render\\", render], __FILE__]);

export { multi as default };
"
`;
